labcontrol/db/process.py

```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
+        value in the column named projname_col_name. Assuming this is true, set
+        the project column value for each non-experimental sample to the value
+        of the project name for the (single) project on the non-experimental
+        sample's plate.
+
+        Parameters
+        ----------
+        input_df: pandas.DataFrame
+            A dataframe containing (at least) a column of plate names (having
+            the column name given in plate_col_name) and a column of project
+            names (having the column name given in projname_col_name)--e.g.,
+            Project Name on prep sheet or sample_proj_name on sample sheet--
+            and one row for each sample (both experimental and non-
+            experimental).  The value in the project name column must be None
+            for control (blank/positive control/etc) samples.
+        plate_col_name: str
+            The name of the column in input_df that contains the name of the
+            plate on which a given sample lies.
+        projname_col_name: str
+            The name of the column in input_df that contains the name of the
+            project associated with the given sample.
+
+        Returns
+        -------
+        input_df: pandas.DataFrame
+            The input dataframe, modified so that the controls have the same
+            (single) project name as the experimental samples on their sample
+            plate.
+
+        Raises
+        ------
+        ValueError
+            If any plate contains experimental samples from more (or fewer)
+            than one project.
+        """
+
+        problem_plate_messages = []
+
+        # create a mask to define all the NON-control rows for this plate
+        non_controls_mask = input_df[projname_col_name].notnull()
+
+        # get all the unique plates in the dataframe
+        unique_plates = input_df[plate_col_name].unique()
+        for curr_unique_plate in unique_plates:
+            # create a mask to define all the rows for this plate
+            plate_mask = input_df[plate_col_name] == curr_unique_plate
+
+            # create a mask to define all the rows for this plate where the
+            # project name is NOT the control value (None)
+            plate_non_controls_mask = plate_mask & non_controls_mask
+
+            # get unique project names for the part of df defined in the mask
+            curr_unique_projnames = (input_df[plate_non_controls_mask]
+                                  [projname_col_name].unique())
+            if len(curr_unique_projnames) != 1:
+                # Note that we don't error out the first time we find a
+                # plate that doesn't meet expectations; instead we continue to
+                # run through all the plates and identify ALL those that don't
+                # meet expectations.  This way the user can correct all of them
+                # at once.
+                curr_err_msg = "Expected one unique value for plate '{0}' " \
```

Comment:
There is a heinous edge case here: it is possible that `len(curr_unique_projnames) == 0`. I don't believe it would break the formation of the error message, but the nature of the error is different.

Reply:
That edge case has already been anticipated and is handled by this logic. The error string recorded in that case looks like, e.g., "Expected one unique value for plate 'Test plate 3' but received 0: ".  I agree that the extra colon at the end is a bit ugly but I don't think it is worth refactoring for an edge case that we keep being told will never occur ... :)